man/udev.xml

```diff
@@ -685,6 +685,62 @@ SUBSYSTEM=="net", OPTIONS="log_level=debug"</programlisting></para>
           </varlistentry>
         </variablelist>
 
+        <para>The order in which expressions in each rule are processed is
+          defined by their type.  First, <varname>LABEL</varname> assignment is
+          performed, then match expressions are processed in the following order:
+          <varname>ACTION</varname>,
+          <varname>DEVPATH</varname>,
+          <varname>KERNEL</varname>,
+          <varname>SYMLINK</varname>,
+          <varname>NAME</varname>,
+          <varname>ENV</varname>,
+          <varname>CONST</varname>,
+          <varname>TAG</varname>,
+          <varname>SUBSYSTEM</varname>,
+          <varname>DRIVER</varname>,
+          <varname>ATTR</varname>,
+          <varname>SYSCTL</varname>,
+          <varname>KERNELS</varname>,
+          <varname>SUBSYSTEMS</varname>,
+          <varname>DRIVERS</varname>,
+          <varname>ATTRS</varname>,
+          <varname>TAGS</varname>,
+          <varname>TEST</varname>,
+          <varname>PROGRAM</varname>,
+          <varname>IMPORT{file}</varname>,
+          <varname>IMPORT{program}</varname>,
+          <varname>IMPORT{builtin}</varname>,
+          <varname>IMPORT{db}</varname>,
+          <varname>IMPORT{cmdline}</varname>,
+          <varname>IMPORT{parent}</varname>,
+          and <varname>RESULT</varname>.
+          If one of these expressions does not match, other expressions in
+          the rule are not processed.  Finally, assignment expressions
+          are executed in the following order:
+          <varname>OPTIONS string_escape</varname>,
+          <varname>OPTIONS db_persist</varname>,
+          <varname>OPTIONS watch</varname>,
+          <varname>OPTIONS link_priority</varname>,
+          <varname>OPTIONS log_level</varname>,
+          <varname>OWNER</varname>,
+          <varname>GROUP</varname>,
+          <varname>MODE</varname>,
+          <varname>TAG</varname>,
+          <varname>OPTIONS static_node</varname>,
+          <varname>SECLABEL</varname>,
+          <varname>ENV</varname>,
+          <varname>NAME</varname>,
+          <varname>SYMLINK</varname>,
+          <varname>ATTR</varname>,
+          <varname>SYSCTL</varname>,
+          <varname>RUN{builtin}</varname>,
+          <varname>RUN{program}</varname>,
+          and <varname>GOTO</varname>.
+          If the order in which expressions are written in the rule is different
+          from the order of processing listed above, these expressions are
+          reordered.
+        </para>
```

Comment:
how does the order actually effectively matter? if this is not explained it's wortheless info

Reply:
There are expressions with side effects, so if a rule contains an expression of this kind, the order is important.
For example, `TEST` might be affected by `PROGRAM` or `IMPORT`, and so on.

Reply:
There are expression types that form groups so that the order withing those group is not important.
For example, all match types before TK_M_PROGRAM form such a group.  These are match expressions without side effects.  It's important that all expressions from this group are processed before other expressions.

Would you suggest documenting these expression groups and the order in which they are processed instead of the current approach?

Reply:
Hmm, documenting those details doesn't seem right. This order of evaluation seems chaotic. If we document this, we are effectively saying that this is something to be relied upon. I don't have a good suggestion what to do, but I'd be wary of merging this.